x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java

```diff
@@ -825,4 +843,128 @@ private BitSet gatherPreAnalysisMetrics(LogicalPlan plan, BitSet b) {
+            for (int i = 0; i < args.size(); i++) {
+                arg = args.get(i);
+                if (arg.resolved() && arg.dataType() == KEYWORD && arg.foldable()) {
+                    if (i < metaData.size()) {
+                        targetDataType = getTargetType(metaData.get(i));
+                    }
+                    if (targetDataType != NULL && targetDataType != UNSUPPORTED) {
+                        Expression e = castStringLiteral(arg, targetDataType);
+                        childrenChanged = true;
+                        newChildren.add(e);
+                        continue;
+                    }
+                }
+                newChildren.add(args.get(i));
+            }
+            return childrenChanged ? f.replaceChildren(newChildren) : f;
+        }
+
+        private static Expression processArithmeticOperation(EsqlArithmeticOperation o) {
+            Expression left = o.left();
+            Expression right = o.right();
+            if (left.resolved() == false || right.resolved() == false) {
+                return o;
+            }
+            List<Expression> newChildren = new ArrayList<>(2);
+            boolean childrenChanged = false;
+            DataType targetDataType = NULL;
+            Expression from = Literal.NULL;
+
+            if (left.dataType() == KEYWORD && left.foldable() && (right.dataType().isNumeric() || right.dataType() == DATETIME)) {
+                targetDataType = right.dataType();
+                from = left;
+            }
+            if (right.dataType() == KEYWORD && right.foldable() && (left.dataType().isNumeric() || left.dataType() == DATETIME)) {
+                targetDataType = left.dataType();
+                from = right;
+            }
+
+            Expression e = from == Literal.NULL ? from : castStringLiteral(from, targetDataType);
+
+            if (e != Literal.NULL) {
+                childrenChanged = true;
+                newChildren.add(from == left ? e : left);
+                newChildren.add(from == right ? e : right);
+            }
+
+            return childrenChanged ? o.replaceChildren(newChildren) : o;
+        }
+
+        private static List<String[]> getMetaData(ScalarFunction f) {
+            List<String[]> paramTypes = new ArrayList<>();
+            var constructors = f.getClass().getConstructors();
+            Constructor<?> constructor = constructors[0]; // no multiple c'tors supported
+            var params = constructor.getParameters();
+            for (int i = 1; i < params.length; i++) { // skipping 1st argument, the source
+                if (Configuration.class.isAssignableFrom(params[i].getType()) == false) {
+                    Param paramInfo = params[i].getAnnotation(Param.class);
+                    String[] type = paramInfo == null ? new String[] { "?" } : paramInfo.type();
+                    paramTypes.add(type);
+                }
+            }
+            return paramTypes;
+        }
```

Comment:
The metadata is static however I'm not sure that's the case for implicit conversions across all functions.

Reply:
> The metadata is static however I'm not sure that's the case for implicit conversions across all functions.

One exception I found during testing is Neg.java extends EsqlScalarFunction, however there is no entry for it in EsqlFunctionRegistry as it is categorized as a unary operator, the rest of EsqlScalarFunction work so far.

Reply:
> Could we grab this from the function registry instead?
> 
> Also! Since we're building the function registry at startup could we build the list of auto-casts into it? Like as a Map<String, Function<Object, Object>> member in `ArgSignature`. You could stick `keyword` and `text` in the map if the types contain `datetime` but not `keyword` or `text`.

@nik9000 I moved the code to decide the target data type into EsqlFunctionRegistry, it is the 5th input to ArgSignature, so that we don't need to copy the code(look into function constructor) into the rule. And I left the code to do the conversion inside the rule, so that it can be shared by ArithmeticOperation and BinaryComparison(in the future).